crates/parcel/src/requests/asset_request.rs

```diff
@@ -0,0 +1,168 @@
+use parcel_core::types::AssetStats;
+use parcel_core::types::Dependency;
+use parcel_core::types::Environment;
+use parcel_core::types::FileType;
+use parcel_filesystem::FileSystemRef;
+
+use crate::plugins::Plugins;
+use crate::plugins::TransformerPipeline;
+use crate::request_tracker::{Request, RequestResult, RunRequestContext, RunRequestError};
+
+/// The AssetRequest runs transformer plugins on discovered Assets.
+/// - Decides which transformer pipeline to run from the input Asset type
+/// - Runs the pipeline in series, switching pipeline if the Asset type changes
+/// - Stores the final Asset source code in the cache, for access in packaging
+/// - Finally, returns the complete Asset and it's discovered Dependencies
+pub struct AssetRequest<'a> {
+  pub env: Arc<Environment>,
+  pub file_path: PathBuf,
+  pub code: Option<String>,
+  pub pipeline: Option<String>,
+  pub side_effects: bool,
+  // TODO: move the following to RunRequestContext
+  pub cache: CacheRef,
+  pub file_system: FileSystemRef,
+  pub plugins: Arc<Plugins<'a>>,
+}
+
+impl<'a> Hash for AssetRequest<'a> {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    // TODO: Just derive this once the contextual params are moved to RunRequestContext
+    self.file_path.hash(state);
+    self.code.hash(state);
+    self.pipeline.hash(state);
+    self.env.hash(state);
+    self.side_effects.hash(state);
+  }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct AssetResult {
+  pub asset: Asset,
+  pub dependencies: Vec<Dependency>,
+}
+
+impl<'a> Request<AssetResult> for AssetRequest<'a> {
+  fn run(
+    &self,
+    request_context: RunRequestContext<AssetResult>,
+  ) -> Result<RequestResult<AssetResult>, RunRequestError> {
+    request_context.report(ReporterEvent::BuildProgress(BuildProgressEvent::Building(
+      AssetBuildEvent {
+        // TODO: Should we try avoid a clone here?
+        file_path: self.file_path.clone(),
+      },
+    )));
+
+    let pipeline = self
+      .plugins
+      .transformers(&self.file_path, self.pipeline.as_deref())?;
+    let asset_type = FileType::from_extension(
+      self
+        .file_path
+        .extension()
+        .and_then(|s| s.to_str())
+        .unwrap_or(""),
+    );
+    let mut transform_ctx = RunTransformContext::new(self.file_system.clone());
+
+    let result = run_pipeline(
+      pipeline,
+      TransformationInput::InitialAsset(InitialAsset {
+        // TODO: Are these clones neccessary?
+        file_path: self.file_path.clone(),
+        code: self.code.clone(),
+        env: self.env.clone(),
+        side_effects: self.side_effects,
+      }),
+      asset_type,
+      &self.plugins,
+      &mut transform_ctx,
+    )?;
+
+    // Write the Asset source code to the cache, this is read later in packaging
+    // TODO: Clarify the correct content key
+    let content_key = result.asset.id().to_string();
+    self
+      .cache
+      .set_blob(&content_key, result.asset.source_code.bytes())?;
```

Comment:
`source_code` is confusing since it's not source anymore it's compiled.

Reply:
Agreed, renamed to `Code`